### DISCUSSION TITLE
Eloquent query conditional using when() and unless()

### DISCUSSION THREAD

**Discussion by ronssij** [laravel/framework#51108] [General]
Would it be nice if we build the `->when()` and `->unless()` for model queries differently or let those method accept Builder class?

From this:

```php
Model::query();
  ->when(
      $isTrue,
      function ($query) {
          $query->whereHas('foos', fn ($query) => $query->where('foo_id', $foo->id)),
      },
      fn ($query) => $query->where('bar', $bar->id)
  );
```

To this:
```php

Model::query();
  ->when(
      $isTrue,
      $query->whereHas('foos', fn ($query) => $query->where('foo_id', $foo->id)),
      $query->where('bar', $bar->id)
  )
  ->get();
  ```

**Comment by henzeb** [comment]
That's not possible... PHP would break over a non existent variable. Even wrapped inside another function that has $query, this would not work as both statements are executed before passed to the when/unless methods...
